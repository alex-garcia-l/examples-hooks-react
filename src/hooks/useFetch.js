import { useEffect, useRef, useState } from "react";

const initialState = {
  data: null,
  loading: true,
  error: null
};

export const useFetch = (url = '') => {

  const [state, setState] = useState(initialState);
  const isMounted = useRef(true);

  useEffect(() => {
    return () => {
      isMounted.current = false;
    }
  }, []);

  useEffect(() => {

    setState(initialState);

    fetch(url)
      .then(resp => resp.json())
      .then(data => {

        if (isMounted.current) {
          setState({
            data,
            loading: false,
            error: null
          });
        }

      })
      .catch((error) => {
        setState({
          data: null,
          loading: false,
          error: 'Error'
        });
      });
  }, [url]);

  return state;
};
