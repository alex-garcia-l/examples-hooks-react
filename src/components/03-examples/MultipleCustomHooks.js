import React from 'react';
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch';

export const MultipleCustomHooks = () => {

  const { counter, increment } = useCounter(1);

  const { loading, data } = useFetch(`https://breakingbadapi.com/api/quotes/${counter}`);
  const { author, quote } = !!data && data[0];

  return (
    <>
      <h1 className="mt-5">Breaking Bad</h1>

      {
        loading
          ?
          (
            <div className="alert alert-info text-center" role="alert">
              Cargando...
            </div>
          )
          :
          (
            <>
              <blockquote className="blockquote p-5 mb-4 bg-light rounded-3 text-end">
                <h2 className="mb-5">{quote}</h2>
                <footer className="blockquote-footer">{author}</footer>
              </blockquote>
              <button
                onClick={() => increment(1)}
                className="btn btn-primary float-end"
              >Siguiente quote</button>
            </>
          )
      }
    </>
  );
};
