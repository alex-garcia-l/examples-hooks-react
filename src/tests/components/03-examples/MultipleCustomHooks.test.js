import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useCounter } from '../../../hooks/useCounter';
import { useFetch } from '../../../hooks/useFetch';
jest.mock('../../../hooks/useCounter');
jest.mock('../../../hooks/useFetch');

describe('Prueba para el componente <MultipleCustomHooks />', () => {

  beforeEach(() => {
    useCounter.mockReturnValue({
      counter: 1,
      increment: () => {}
    });
  })

  test('debe mostrarse correctamente', () => {

    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe mostrar informaciÃ³n', () => {

    const data = {
      author: 'Developer',
      quote: 'A quote'
    };

    useFetch.mockReturnValue({
      data: [data],
      loading: false,
      error: null
    });

    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('h2.mb-5').text().trim()).toBe(data.quote);
    expect(wrapper.find('footer').text().trim()).toBe(data.author);
  });
});
