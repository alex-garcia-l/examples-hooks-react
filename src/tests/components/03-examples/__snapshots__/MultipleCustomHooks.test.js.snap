// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba para el componente <MultipleCustomHooks /> debe mostrar informaci√≥n 1`] = `
<Fragment>
  <h1
    className="mt-5"
  >
    Breaking Bad 
    <span
      className="fs-3"
    >
      (Query API)
    </span>
  </h1>
  <hr />
  <blockquote
    className="blockquote p-5 mb-4 bg-light rounded-3 text-end"
  >
    <h2
      className="mb-5"
    >
      A quote
    </h2>
    <footer
      className="blockquote-footer"
    >
      Developer
    </footer>
  </blockquote>
  <button
    className="btn btn-primary float-end"
    onClick={[Function]}
  >
    Siguiente quote
  </button>
</Fragment>
`;

exports[`Prueba para el componente <MultipleCustomHooks /> debe mostrarse correctamente 1`] = `
<Fragment>
  <h1
    className="mt-5"
  >
    Breaking Bad 
    <span
      className="fs-3"
    >
      (Query API)
    </span>
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
    role="alert"
  >
    Cargando...
  </div>
</Fragment>
`;
